### PR TITLE
Create Abstractions library for Async and IIncludable Extensions

### PR DESCRIPTION
**Pull request by rdefreitas**
Create Abstractions library for Async and IIncludable Extensions
Per proposal #6085

- Create project/package Microsoft.EntityFrameworkCore.Abstractions
- Microsoft.EntityFrameworkCore now takes a dependency on ...Abstractions
- Add IEntityQueryProvider marker interface to aide in determining if a query is provided by Entity Framework.
- Async Linq extensions fallback to synchronous counterparts
- move async linq methods into AsyncQueryableExtensions with non-extensions fallback in EntityFrameworkQueryableExtensions (to reduce breaking changes)
- move AsyncQueryableExtensions to System.Linq namespace (per Guiding Principles)

### CODE REVIEW

**Comment by rdefreitas** [comment]
phew, those functional tests with the string-based build references gave me a nice headache tracking down the issue (particularly when the build takes about 20 minutes).

Reopening this PR now that I tracked down what broke the build... ;-)

**Comment by rowanmiller** [comment]
cc @natemcmaster to comment on the AppVeyor issue

**Comment by natemcmaster** [comment]
I triggered a rebuild manually.

File: src/Microsoft.EntityFrameworkCore.Abstractions/Properties/CoreStrings.resx [hunk 1]
```diff
@@ -565,7 +565,7 @@
     <value>Cannot create a DbSet for '{typeName}' because this type is not included in the model for the context.</value>
   </data>
   <data name="AmbiguousOneToOneRelationship" xml:space="preserve">
-    <value>The child/dependent side could not be determined for the one-to-one relationship that was detected between '{dependentToPrincipalNavigationSpecification}' and '{principalToDependentNavigationSpecification}'. To identify the child/dependent side of the relationship, configure the foreign key property. See http://go.microsoft.com/fwlink/?LinkId=724062 for more details.</value>
+    <value>The child/dependent side could not be determined for the one-to-one relationship that was detected between '{dependentToPrincipalNavigationSpecification}' and '{principalToDependentNavigationSpecification}'. To identify the child/dependent side of the relationship, configure the foreign key property. See http://go.microsoft.com/fwlink/?LinkId=724062 for more details.\</value>
```
Comment (divega) [inline src/Microsoft.EntityFrameworkCore.Abstractions/Properties/CoreStrings.resx]: Typo?

File: src/Microsoft.EntityFrameworkCore/EntityFrameworkQueryableExtensions.cs [hunk 2]
```diff
@@ -325,18 +277,13 @@ public static class EntityFrameworkQueryableExtensions
         ///     <paramref name="predicate" />.
         /// </returns>
         public static Task<TSource> FirstAsync<TSource>(
-            [NotNull] this IQueryable<TSource> source,
+            [NotNull] IQueryable<TSource> source,
             [NotNull] Expression<Func<TSource, bool>> predicate,
-            CancellationToken cancellationToken = default(CancellationToken))
+            CancellationToken cancellationToken = new CancellationToken())
```
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore/EntityFrameworkQueryableExtensions.cs]: Why are these changing?

File: src/Microsoft.EntityFrameworkCore/EntityFrameworkQueryableExtensions.cs [hunk 3]
```diff
@@ -83,18 +75,13 @@ public static class EntityFrameworkQueryableExtensions
         ///     predicate; otherwise, <c>false</c>.
         /// </returns>
         public static Task<bool> AnyAsync<TSource>(
-            [NotNull] this IQueryable<TSource> source,
+            [NotNull] IQueryable<TSource> source,
             [NotNull] Expression<Func<TSource, bool>> predicate,
-            CancellationToken cancellationToken = default(CancellationToken))
+            CancellationToken cancellationToken = new CancellationToken())
```
Comment (ilmax) [inline src/Microsoft.EntityFrameworkCore/EntityFrameworkQueryableExtensions.cs]: Why there are mixed `new CancellationToken()` and `default(CancellationToken)` ?

File: src/Microsoft.EntityFrameworkCore.Abstractions/Extensions/Internal/QueryableExtensions.cs [hunk 4]
```diff
@@ -21,6 +23,20 @@ public static class QueryableExtensions
         ///     This API supports the Entity Framework Core infrastructure and is not intended to be used 
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
+        public static bool IsAsyncEnumerable<TSource>([NotNull] this IQueryable<TSource> source)
+        {
+            Check.NotNull(source, nameof(source));
+
+            var enumerable = source as IAsyncEnumerable<TSource>;
```
Comment (ilmax) [inline src/Microsoft.EntityFrameworkCore.Abstractions/Extensions/Internal/QueryableExtensions.cs]: can be simplified with `return source is IAsyncEnumerable<TSource>`

File: src/Microsoft.EntityFrameworkCore/EntityFrameworkQueryableExtensions.cs (hunk 3 above)
Comment (rdefreitas) [inline src/Microsoft.EntityFrameworkCore/EntityFrameworkQueryableExtensions.cs]: In most cases it's default... apparently I got one mixed somewhere along the way, and it may have been an auto formatting changed based on resharper settings on my machine. I'll go back and fix these up.

File: src/Microsoft.EntityFrameworkCore.InMemory/project.json [hunk 5]
```diff
@@ -23,6 +23,7 @@
     }
   },
   "dependencies": {
+    "Microsoft.EntityFrameworkCore.Abstractions": "1.1.0-*",
```
Comment (rdefreitas) [inline src/Microsoft.EntityFrameworkCore.InMemory/project.json]: added to support a minor defect in a related project that was intermixing EF core with 1.0 RTM libraries. This may be redundant outside of mixed-library scenarios such as that, but it doesn't hurt anything to leave it in place.

File: src/Microsoft.EntityFrameworkCore.Relational/project.json [hunk 6]
```diff
@@ -25,6 +25,7 @@
     }
   },
   "dependencies": {
+    "Microsoft.EntityFrameworkCore.Abstractions": "1.1.0-*",
```
Comment (rdefreitas) [inline src/Microsoft.EntityFrameworkCore.Relational/project.json]: added to support a minor defect in a related project that was intermixing EF core with 1.0 RTM libraries. This may be redundant outside of mixed-library scenarios such as that, but it doesn't hurt anything to leave it in place.

Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Relational/project.json]: @rdefreitas Starting to look at this. First couple of things:

1) Microsoft.EntityFrameworkCore.csproj no longer builds. Looks like some shared files got removed.
2) CoreStrings.resx got moved. If we need resources in Abstractions, then split them out into an AbstractionsStrings.resx and link the Resources.tt file

Thanks!

File: src/Microsoft.EntityFrameworkCore.Abstractions/Properties/InternalsVisibleTo.cs [hunk 7]
```diff
@@ -0,0 +1,11 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("Microsoft.EntityFrameworkCore, PublicKey=0024000004800000940000000602000000240000525341310004000001000100f33a29044fa9d740c9b3213a93e57c84b472c84e0b8a0e1ae48e67a9f8f6de9d5f7f3d52ac23e48ac51801f1dc950abe901da34d2a9e3baadb141a17c77ef3c565dd5ee5054b91cf63bb3c6ab83f72ab3aafe93d0fc3c2348b764fafb0b1c0733de51459aeab46580384bf9d74c4e28164b7cde247f891ba07891c9d872ad2bb")]
```
Comment (divega) [inline src/Microsoft.EntityFrameworkCore.Abstractions/Properties/InternalsVisibleTo.cs]: Why is this necessary? In principle it doesn't seem right.

File: src/Microsoft.EntityFrameworkCore.Abstractions/Query/Internal/IEntityQueryProvider.cs [hunk 8]
```diff
@@ -0,0 +1,40 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Linq;
+using System.Linq.Expressions;
+
+// ReSharper disable once CheckNamespace
+namespace Microsoft.EntityFrameworkCore.Query.Internal
```
Comment (divega) [inline src/Microsoft.EntityFrameworkCore.Abstractions/Query/Internal/IEntityQueryProvider.cs]: It seems weird to have internal surface in an abstractions package.

**Comment by rdefreitas** [comment]
@natemcmaster it seems this will need another manual kickoff before it can be reviewed again.
> Build execution time has reached the maximum allowed time for your plan (60 minutes).

Thank you, sir.

File: src/Microsoft.EntityFrameworkCore.Abstractions/Query/Internal/IEntityQueryProvider.cs (hunk 8 above)
Comment (rdefreitas) [inline src/Microsoft.EntityFrameworkCore.Abstractions/Query/Internal/IEntityQueryProvider.cs]: @divega I has split out the small necessary components from the EF Core project into the Abstractions project, and added a dependency on the abstractions in the Ef library, with a focus on minimizing breaking changes..

With the exception of classes whose namespaces were intentionally changed, I tried to keep the same namespaces intact whenever possible, particular in the case of highly related components.

Considering that all of the Internal namespaces give warnings in the code docs for outside parties not to rely on them, I suppose it's safe to change them, and in doing so, simplify the library a bit.

IEntityQueryProvider is a helper interface that was introduced in place of EntityQueryProvider and used in the extensions for IIncludableQueryable, as the EntityQueryProvider type is no longer accessible to those extensions.  Would you propose I eliminate any "Internal" namespaces and only preserve old namespaces if they would have resulted in breaking changes?

I could move everything into the Microsoft.EntityFrameCore.Abstractions namespace unless there is a good case for why it lead to breaking changes.

Comment (divega) [inline src/Microsoft.EntityFrameworkCore.Abstractions/Query/Internal/IEntityQueryProvider.cs]: @rdefreitas In general the bar for moving internal types to public surface his quite high, so no, I am not suggesting that we move anything. My main point is that carrying internal types in an Abstractions packages smells. It may be a symptom that what you are creating is not really an abstractions package, that the factoring needs improvement or - like with any smell - it maybe ok for some reason (although I would be suprised). Anyway, I will let you and @anpete figure that out :smile:

File: src/Microsoft.EntityFrameworkCore.Abstractions/Properties/InternalsVisibleTo.cs (hunk 7 above)
Comment (rdefreitas) [inline src/Microsoft.EntityFrameworkCore.Abstractions/Properties/InternalsVisibleTo.cs]: I agree that it feels wrong...

It was necessary to move CoreStrings in Abstractions so that the embedded resources could be accessed there, but oddly the shared utility classes, particularly Check, were encountering type collision issues when they existed in both libraries.

In light of the collision and in the interest of minimizing future support, I did not wish to break out CoreStrings and Check to include a different internal version in each of the 2 libraries. allowing EFCore to reference those from Abstractions.

I can go back to that issue this weekend and see if maybe there was something overlooked that led to that type collision during builds, although nothing comes to mind off-hand. If you have any suggestions I'm open to them :)

File: src/Microsoft.EntityFrameworkCore.Abstractions/Query/Internal/IEntityQueryProvider.cs (hunk 8 above)
Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Abstractions/Query/Internal/IEntityQueryProvider.cs]: @rdefreitas Change the Include stuff to test for IAsyncQueryProvider.

Comment (rdefreitas) [inline src/Microsoft.EntityFrameworkCore.Abstractions/Query/Internal/IEntityQueryProvider.cs]: To be honest, I think these extensions, with the exception of IIncludableQueryable would feel more natural as part of System.Interactive.Async in [Reactive-Extensions/Rx.NET](https://github.com/Reactive-Extensions/Rx.NET), but then that would leave the questions of the best way to hook up special handling for EF's EntityQueryProvider, where the Includable extensions would live, and of course the overall issues of API coupling across disparate projects.

The idea here was to allow an indispensable set of Extension methods from Entity Framework to be utilized in service libraries that do not directly reference the bulk of EFCore, without the need to introduce proxy/wrapper classes which easily can get out of hand.

File: src/Microsoft.EntityFrameworkCore.Abstractions/Properties/InternalsVisibleTo.cs (hunk 7 above)
Comment (rdefreitas) [inline src/Microsoft.EntityFrameworkCore.Abstractions/Properties/InternalsVisibleTo.cs]: @anpete this covers the reason behind the moving of CoreStrings.resx, although I did not encounter the build issues you're speaking of regarding EntityFrameworkCore.csproj ... were you building that in isolation from the solution and is that a scenario that needs to be supported?

Comment (anpete) [inline src/Microsoft.EntityFrameworkCore.Abstractions/Properties/InternalsVisibleTo.cs]: No, opened EntityFramework.sln and hit build.

**Comment by rdefreitas** [comment]
@anpete based on feedback, here's the changes I'll be looking at this weekend... if you have any other suggestions before I dive back please let me know.

- [ ] Move CoreStrings back to EntityFrameworkCore.
- [ ] Remove InternalsVisibleTo
- [ ] Remove reference to shared NotNull checks - add subset version for the Abstractions library only, utilizing an Abstractions resource file for the error messages that should be utilized by the Abstractions lib.
- [ ] Split AsyncQueryableExtensions and IncludableExtensions (note, since the library is very focused on providing extension methods, should we break this out into partial static classes as opposed to the regions that exist now?)
- [ ] Drink more coffee :)

**Comment by anpete** [comment]
@rdefreitas Looks good. To address the internal IAsyncQueryProvider issue, we are thinking that we can have IEntityQueryProvider in Abstractions and put the async query methods on it. So, IEntityQueryProvider can be used from both the async and Include extension methods. Sound reasonable?

Re the Check library, just link the whole file, like we do for everything else.

Fine to use partials to better organize the extension methods.

**Comment by rdefreitas** [comment]
@anpete, that all seems fine with the exception of the Check library, since it references the CoreStrings resources from EntityFrameworkCore and this library is independent of the EFCore base lib.

Also, do I need to install a separate version of the Visual Studio tooling?  The .sln file never worked for me in VS, only the -xproj.sln

**Comment by anpete** [comment]
@rdefreitas Oh, I see the problem now. Let's just lift the Check strings to the Abstractions resource file.

**Comment by rdefreitas** [comment]
@anpete I ran into personal delays on my end in being able to get back to this PR, so in the interest of still having a worthwhile enhancement for 1.1.0 depending on release date, I've streamlined a separate pull request to just provide the synchronous fallback to the existing methods and will come back to this another time.
